common/src/main/kotlin/streams/service/sink/strategy/PatternConfiguration.kt

```diff
@@ -0,0 +1,116 @@
+        PatternConfigurationType.ALL -> emptyList()
+    }
+}
+
+interface PatternConfiguration
+
+data class NodePatternConfiguration(val keys: Set<String>, val type: PatternConfigurationType,
+                                    val labels: List<String>, val properties: List<String>): PatternConfiguration {
+    companion object {
+
+        private val nodePatternConfigured = "^(\\w+\\s*(?::\\s*(?:\\w+)\\s*)*)\\s*(?:\\{\\s*(-?[\\w|!|\\.]+\\s*(?:,\\s*-?[!|\\w|\\*|\\.]+\\s*)*)\\})?\$".toRegex()
+        fun parse(pattern: String): NodePatternConfiguration {
+            val matcher = nodePatternConfigured.matchEntire(pattern)
+            if (matcher == null) {
+                throw IllegalArgumentException("The pattern $pattern is invalid")
+            } else {
+                val labels = matcher.groupValues[1].split(LABEL_SEPARATOR).map { it.quote() }
+                val allProperties = matcher.groupValues[2].split(PROPERTIES_SEPARATOR)
+                val keys = allProperties
+                        .filter { it.startsWith(ID_PREFIX) }
+                        .map { it.trim().substring(1) }.toSet()
+                if (keys.isEmpty()) {
+                    throw IllegalArgumentException("The pattern $pattern must contains at lest one key")
+                }
+                val properties = allProperties.filter { !it.startsWith(ID_PREFIX) }
+                val type = getPatternConfiguredType(properties)
+                isHomogeneousPattern(type, properties, pattern)
+                val cleanedProperties = cleanProperties(type, properties)
+
+                return NodePatternConfiguration(keys = keys, type = type,
+                        labels = labels, properties = cleanedProperties)
+            }
+        }
+    }
+}
+
+
+data class RelationshipPatternConfiguration(val start: NodePatternConfiguration, val end: NodePatternConfiguration,
+                                            val relType: String, val type: PatternConfigurationType,
+                                            val properties: List<String>): PatternConfiguration {
+    companion object {
+
+        // we don't allow ALL for start/end nodes in rels
+        // it's public for testing purpose
+        fun getNodeConf(pattern: String): NodePatternConfiguration {
+            val start = NodePatternConfiguration.parse(pattern)
+            return if (start.type == PatternConfigurationType.ALL) {
+                NodePatternConfiguration(keys = start.keys, type = PatternConfigurationType.INCLUDE,
+                        labels = start.labels, properties = start.properties)
+            } else {
+                start
+            }
+        }
+
+        // (:Source{!id})-[:REL_TYPE{foo, -bar}]->(:Target{!targetId})
+        private val relationshipPatternConfigured = "^\\((.*?)\\)-\\[(?::)([\\w|\\_]+)(\\{\\s*(-?[\\w|\\*|\\.]+\\s*(?:,\\s*-?[\\w|\\*|\\.]+\\s*)*)\\})?\\]->\\((.*?)\\)\$".toRegex()
```

Comment:
you can use """ strings to remove the double backslashes
you don't need the `|` in the square brackets `[\\w|\\*|\\.]` [a-z0-9]

Reply:
Ok